docs/csharp/roslyn-sdk/incremental-source-generators/overview.md

```diff
@@ -0,0 +1,158 @@
+---
+# Roslyn Incremental Source Generators Overview
+
+Source generators are run by the compiler to add C# or VB code to the compilation. Roslyn *incremental source generators* are defined as a series of steps called by the compiler which allows the compiler to shortcut if the results of any step in its cache is unchanged. Incremental generators are also granular and manage the cache independently for each item in a collection. The compiler can also cancel after each step. These features allow incremental generators to minimize the work performed and provide a pattern that is maintainable and testable.
+
+Source generators run during both design time builds and [[full]] builds. This allows them to extend user code with new code that adds classes, methods and properties that immediately affect the users experience via features like IntelliSense. It also means that generators run very often, possibly on every keystroke.
+
+Most programmers will not need to write source generators, but will take advantage of those written by others. If you do write a generator, follow the [Performance Guidelines]() and be [tested for performance]() because your generator can adversely affect the design time performance of editors such as Visual Studio. Source generators are new and we are evolving guidance on banned APIs, so feel free to reach out at [[@Jaredpa and @chsienki where?]] if you want to check the impact of a technique.
+
+> [!CAUTION]
+> We will disable generators during the design time build if they are slow or we analysis indicates that they do not follow guidelines.
+
+This article assumes that you understand the basics of the [Roslyn compiler](), [syntax trees](), [compilations](), [compilation](), [cancellation tokens]() and the [difference between value and reference equality](). It will also help to have an understanding of functional techniques using delegates such as lambda expressions.
+
+## How Roslyn generators work
+
+When the Roslyn compiler runs without generation, it reads user code and builds a set of syntax trees, generally one syntax tree per file. Based on this, the compiler builds a compilation and emits [IL]() based on the compilation:
+
+```mermaid
+flowchart LR
+    SyntaxTrees-->Compilation-->IL
+```
+
+When generators are present, the Roslyn compiler inserts an extra step between creating the compilation and emitting the IL which runs the generators and adds any syntax trees they create. It then rebuilds the compilation before emitting the IL:
+
+```mermaid
+flowchart LR
+    SyntaxTrees-->Comp1[Compilation]-->Generators-->More[More SyntaxTrees]-->Comp2[Compilation]-->IL
+    SyntaxTrees-->Generators
+    SyntaxTrees-->Comp2
+```
+
+Syntax trees are a full representation of the code the user wrote. The compilation includes several things that may be interersting to your generator, including the semantic model.
+
+In some cases additional information is needed, perhaps from a configuration file:
+
+```mermaid
```

Comment:
Unfortunately, the docs.microsoft.com platform doesn't natively support mermaid. So, if you're expecting this to render as an actual diagram, you'd be better off making this an image from the rendering. To do so, I usually use the https://mermaid.live/ tool.

Reply:
Thanks! This is an early draft to get the dev team's review and I'll change it, probably with the tool you suggest. For now, I want to keep it as easy as possible to review (I am also inlining all the code, which will change later).